[#1] Layout, Pages Router Setup, and Login Page

kleekich21 opened the pull request:
## 상세
- [X] Pages Router 셋업
- [x] Common Layout with Navbar
- [x] Pages Boiler Plate files
- [x] Add React Hook Form to Login Page
- [x] LocalStorage, SessionStorage 사용 추상화
- [x] User Data 관리를 위해 react-query와 LocalStorage 사용
- [x] Handling UnhandledRejectionError with Custom Error
- [x] Tests

## 이번 PR에서 중점적으로 봐주셨으면 하는 내용
- Pages  Router 적용 내용
- 모든 페이지에서 사용되는 Layout 적용 내용
- Login Page React Hook Form 적용 내용
- Navbar에서 로그인 한 유저 email 노출 방법

## 참고
- Boiler Plate만 추가된 페이지는 테스트 코드 작성을 아직 하지 않았습니다.

kleekich21 committed:
feat: WIP

kleekich21 committed:
chore: update README

kleekich21 committed:
chore: directory structure has changed, user email is displayed when loggedin

kleekich21 committed:
feat: Apply React Hook Form to Login page and handles login and signup with supabase client

kleekich21 committed:
feat: Add toast to Login page

kleekich21 committed:
test: add tests and update git actions

kleekich21 committed:
test: cleanup

kleekich21 committed:
chore: update github actions file for running test command

kleekich21 committed:
chore: add vitest code coverage library

kleekich21 committed:
test: remove duplicate test code

f-lab-cadiz commented on src/components/Common/Layout/Navbar/index.tsx:
```diff
@@ -0,0 +1,67 @@
+export default function Navbar() {
+  const pathname = usePathname();
+
+  const [email, setEmail] = useState<string | undefined>("Login");
+
+  useEffect(() => {
+    const revalidateUser = async () => {
+      try {
+        const supabase = createClient();
+        const {
+          data: { user },
+        } = await supabase.auth.getUser();
+
+        setEmail(user?.email);
+      } catch (error) {
+        console.error("Error fetching user data:", error);
+      }
+    };
+    revalidateUser();
+  }, [email]);
```
이 부분이 react-quuery로 되어 있으면 좋을 것 같네요

kleekich21 replied:
auth 관련 하여 리펙토링하여 더이상 코드가 존재하지 않습니다.

kleekich21 replied:
CSR로 변환 뒤 useQuery로 적용 하였습니다.

f-lab-cadiz commented on src/components/Common/Layout/Navbar/index.tsx:
```diff
@@ -0,0 +1,67 @@
+  useEffect(() => {
+    const revalidateUser = async () => {
+      try {
+        const supabase = createClient();
```
혹시 매번 client를 만들어주는게 권장되는 방식인가요? (잘 몰라서요)

kleekich21 replied:
네. 필요할 때마다 client생성해도 무방합니다. [Official Document](https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=pages#:~:text=Component%20%2D%20Runs%20on%20the%20client.%20Reads%20cookies%20from%20browser%20storage.%20Behind%20the%20scenes%2C%20createBrowserClient%20reuses%20the%20same%20client%20instance%20if%20called%20multiple%20times%2C%20so%20don%27t%20worry%20about%20deduplicating%20the%20client%20yourself.)의 설명에 의하면 createBrowserClient는 매번 같은 클라이언트 인스턴스를 재사용한다고 하네요. 중복 생성을 따로 핸들링 할 필요가 없다고 하네요.

링크는 아래 내용과 같습니다.

> Component - Runs on the client. Reads cookies from browser storage. Behind the scenes, createBrowserClient reuses the same client instance if called multiple times, so don't worry about deduplicating the client yourself.

f-lab-cadiz commented on src/components/LoginPage/index.tsx:
```diff
@@ -0,0 +1,125 @@
+  password: string;
+}
+
+export default function LoginPage() {
```
회원가입, 로그인 시 서버와 통신할 때 적절한 Loading 처리가 있으면 좋을 것 같아요 (중복 요청, 클릭 방지)

kleekich21 replied:
관련 로직 추가하였습니다.

f-lab-cadiz commented on src/components/LoginPage/index.tsx:
```diff
@@ -0,0 +1,125 @@
+        `로그인에 실패하였습니다. 다시 시도해주세요. 이유: ${error.message}`
+      );
+    } else {
+      router.push("/games");
```
요러면 뒤로가기 했을 때 다시 로그인 페이지로 오지 않나용?

kleekich21 replied:
네. 로그인 성공하면 login page 접근을 막는것이 맞을 것 같네요. history.replaceState() 함수와 같은 기능을 하는 [nextJS의 router.replace()](https://nextjs.org/docs/pages/api-reference/functions/use-router#routerreplace)로 처리하도록 하겠습니다. Login Page에 접근할 때, login 여부를 확인하고 '/'으로 redirection하는 것이 나을것 같네요.

f-lab-cadiz commented on src/components/LoginPage/index.tsx:
```diff
@@ -0,0 +1,125 @@
+  async function signUp({ email, password }: LoginFormInput) {
+    const { error } = await supabase.auth.signUp({ email, password });
+    if (error) {
+      let errorMsg =
```
요건 왜 let 인가요?

kleekich21 replied:
const가 맞습니다. 감사합니다. :)

kleekich21 committed:
feat: display login status on Navbar, and control access to login page depending on login status

kleekich21 committed:
chore: add react-query setup

kleekich21 committed:
test: add test codes

kleekich21 committed:
test: update for failing github actions

kleekich21 committed:
test: update tests

kleekich21 committed:
test: update tests

kleekich21 committed:
test: update tests

kleekich21 committed:
feat: Loading for login and signup

f-lab-cadiz commented on src/components/LoginPage/index.tsx:
```diff
@@ -0,0 +1,136 @@
+
+  async function logIn({ email, password }: LoginFormInput) {
+    setIsLoading(true);
+    const { error } = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    });
+    setIsLoading(false);
```
signInWithPassword에서 오류가 발생하면 UI가 무한 로딩에 빠질 수 있을 것 같아요

kleekich21 replied:
현수님 말씀이 맞네요. 감사합니다. try, catch, finally 문으로 처리하도록 하겠습니다.

f-lab-cadiz commented on src/components/LoginPage/index.tsx:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  async function signUp({ email, password }: LoginFormInput) {
+    setIsLoading(true);
+    const { error } = await supabase.auth.signUp({ email, password });
+    setIsLoading(false);
```
여기도 무한 로딩에 빠질 수 있을 것 같아요

kleekich21 replied:
여기도 동일하게 처리하겠습니다.

f-lab-cadiz commented on src/queries/auth.ts:
```diff
@@ -0,0 +1,33 @@
+    } = await supabase.auth.getUser();
+    return user;
+  } catch (error) {
+    console.error("Error fetching user data:", error);
```
catch 문에서 아무 동작도 안하는 것 같은데 더 좋은 방법은 없을까요?

kleekich21 replied:
toast를 사용해 유저에게 에러 메시지를 노출하는 것이 좋을 것 같네요.

kleekich21 committed:
chore: update README

kleekich21 committed:
feat: handle infinite loading case with try and catch

kleekich21 committed:
refactor: Login Page, Login Status using react-query

kleekich21 committed:
feat: apply feedback

- API 통신하는 함수들 모아놓기
- 로딩 관련 로직 추상화 하기
- 에러 핸들링 추상화 하기
    1. unhandledRejection Event 처리
- 전역 상태 잘 관리하기 (localStorage, use-query initialValue 활용)
    1. localStorage 잘 추상화 하기
        a. localStorage 없을 때 잘 동작하는가 (서버, Node 환경)
        b. storage 선택이 가능한가 (sessionStorage)
        c. 값을 가져오거나 설정할 때 마다 직접 키에 접근하지 않고 편하게 사용할 수 있는 인터페이스가 제공되는가

kleekich21 committed:
refactor: error handling

kleekich21 committed:
chore: cleanup

f-lab-cadiz commented on src/components/ScorePage/index.tsx:
```diff
@@ -1,14 +1,10 @@
   return (
     <Layout>
-      <h1>Hello, {user.email || "user"}! This is Your Score Page</h1>;
+      <h1>Hello, {getUserEmail() || "user"}! This is Your Score Page</h1>;
```
```suggestion
      <h1>Hello, {getUserEmail() ?? "user"}! This is Your Score Page</h1>;
```

f-lab-cadiz replied:
ESLint 룰을 최대한 strict하게 가져가보면 좋을 것 같아요~
https://typescript-eslint.io/rules/strict-boolean-expressions/ 여기서 잡아주거든요

kleekich21 replied:
Next에서 기본적으로 typescript관련 eslint 룰을 설정해주는 걸로 알고 설치했는데, 뭔가 document에 오류가 있어 보이네요. 
https://github.com/vercel/next.js/issues/69184
메뉴얼로 추가하는 방법을 찾아봐야겠네요.

f-lab-cadiz commented on src/pages/_app.tsx:
```diff
@@ -15,18 +15,52 @@ const queryClient = new QueryClient({
+   * @returns 없음 (void)
+   */
+  const handleUnhandledRejection = (event: PromiseRejectionEvent) => {
+    console.warn(`UNHANDLED PROMISE REJECTION: ${event.reason}`);
+    router.push("/error");
+    toast.error(`에러가 발생하였습니다. 이유: ${event.reason}`);
```
👍 

근데 에러났을 때 UX상 페이지가 이동되면 불편할 것 같아요

f-lab-cadiz commented on src/pages/games/index.tsx:
```diff
@@ -1 +1,7 @@
-export { default } from "@/components/GamesPage";
+import dynamic from "next/dynamic";
+
+const GamesPage = dynamic(() => import("@/components/GamesPage"), {
+  ssr: false,
+});
```
_app에서 적용하면 페이지별로 이렇게 적용하지 않아도 될텐데 그쪽으로 바뀌면 좋을 것 같아요

kleekich21 replied:
코드가 중복되는 것은 저도 인지했습니다. 페이지별로 SSR을 off시킨 이유는, 랜딩페이지도 영향을 받아 SEO에 좋지 않을 것 같은데요. 개인 프로젝트라 크게 문제는 없겠지만, 프로덕션에서는 고려해봐야할 문제겠네요. 관련하여 동적으로 APP 컴포넌트를 로딩하는 방식으로 리펙토링 진행하였습니다.

f-lab-cadiz commented on src/utils/storage/auth.ts:
```diff
@@ -0,0 +1,25 @@
+// 스토리지 타입을 명시하기 위한 enum
```
```
export const storage = createStorage()
export const emailStorage = storage.create('email')

emailStorage.set('foo@bar.com')
emailStorage.get()
emailStorage.delete()
```

이런 느낌으로 추상화 레이어가 하나 있으면 좋을 것 같아요~

kleekich21 replied:
이렇게 추상화 할 경우, 한번만 생성해서 앱 전체에서 재사용이 가능해야할 것 같습니다. 싱글톤 객체를 생성해서 공유하도록 하겠습니다.

kleekich21 committed:
chore: typescript-eslint manual setup(the way create-next-app setup eslint for typescript has some issue - https://github.com/vercel/next.js/issues/69184)

kleekich21 committed:
chore: cleanup

kleekich21 committed:
feat: apply CSR on App component(originally CSR applied to page component)

kleekich21 committed:
refactor: no more routing to error page for unhandledrejection error

kleekich21 committed:
feature: localStorage WIP

kleekich21 committed:
refactor: add abstract layer for storage util. Use singleton pattern to avoid duplicating storage object.

kleekich21 committed:
feat: add serialize and deserialize functions to store various data types in localStorage

kleekich21 committed:
refactor: using custom Error to handle login and signup error

kleekich21 committed:
test: replace vitest with jest

f-lab-cadiz commented on src/utils/storage/createStorage.ts:
```diff
@@ -0,0 +1,53 @@
+  return {
+    create: function (key: string) {
+      return {
+        set: function (value: any) {
+          storage.setItem(key, serializeForStorage(value));
+        },
+        get: function () {
```
이쪽이 지금 타입이 any로 되고 있어서 Generic 받는 방향으로 바뀌어야 할 것 같아요

kleekich21 replied:
반영하였습니다.

f-lab-cadiz commented on src/queries/auth.ts:
```diff
@@ -0,0 +1,22 @@
+  queryKey: ["authenticate"],
+  queryFn: authenticateUser,
+  // staleTime: 300 * 1000, // Supabase session lasts for 5 minutes ~ 1 hour
+  initialData: emailStorage.get(),
```
placeholderData 랑 차이도 한 번 찾아봐주실 수 있나요 ㅎㅎ

kleekich21 replied:
placeholderData가 더 어울려보이네요. 저희는 현재 localStorage를 활용 중이니 첫 랜더링 시 필요한 데이터가 이미 있는 상태이고, cache를 사용할 필요가 없겠네요. placeholderData로 변경하는게 좋을 것 같습니다. 감사합니다.

https://tanstack.com/query/latest/docs/framework/react/guides/placeholder-query-data

f-lab-cadiz commented on src/requests/auth.ts:
```diff
@@ -0,0 +1,41 @@
+    throw new LoginError(error.message);
+  }
+
+  return handleUserEmail(data.user?.email) ? data : undefined;
```
User 객체 내에 쓰일 다른 값들도 추후에 쓰일 수도 있으니 다 저장해두는 건 어떨까요?

kleekich21 replied:
그럼 emailStorage를 생성하는 것이 아닌, UserStorage를 생성하고 통합적으로 관리하는게 나아보이는데 어떻게 생각하시나요?

f-lab-cadiz replied:
좋아요!

kleekich21 replied:
반영하였습니다.

f-lab-cadiz commented on src/requests/auth.ts:
```diff
@@ -0,0 +1,41 @@
+  return handleUserEmail(data.user?.email) ? data : undefined;
+}
+
+export async function requestToSignUp(email: string, password: string) {
+  const { data, error } = await supabase.auth.signUp({ email, password });
+
+  if (error) {
+    throw new SignUpError(error.message);
+  }
```
supabase 쓰는 모든 곳에 이런식으로 처리해줘야 할 것 같아요. 고차함수를 하나 만들어서 성공적인 응답에 관련된 코드만 작성할 수 있도록 개선해보면 좋을 것 같아요.

kleekich21 replied:
이런식으로 처리해야한다는 말씀은 CustomError를 사용해서 에러를 처리해야한다는 말씀이시죠?

고차 함수 관련해서는 좀더 자세하게 설명해주실 수 있으실까요?

f-lab-cadiz replied:
supabase API 쓰는 모든 곳에서 다 에러가 있는지 체크하고 throw를 해줘야 할텐데 이걸 한 번 추상화하면 어떨까 싶었어요

kleekich21 replied:
authenticateUser 함수 같은 경우는 throw로 처리하면 안되는 케이스입니다. supabase API 쓰는 모든 곳에 에러 처리는 한번에 가능한데, 따로 authenticateUser 케이스만 분기하기도 좀 애매하네요. 관련하여 구두로 설명 드리도록 하겠습니다.

kleekich21 committed:
refactor: avoid any and use generictype

kleekich21 committed:
refactor: remove emailStorage and use userStorage

kleekich21 committed:
refactor: serialize and deserialize

kleekich21 committed:
refactor: add another abstract layer to handle supabase auth error with CustomError

f-lab-cadiz commented on src/requests/auth.ts:
```diff
@@ -0,0 +1,54 @@
+}
+
+async function handleSupabaseRequest(
+  fn: () => Promise<SupabaseAuthResponse>
```
(마이너) 콜백을 안받고 바로 Promise를 받으면 좀 더 간결해질 것 같긴 하네요!

f-lab-cadiz reviewed (approved):
👍

kleekich21 committed:
refactor: simplify funcion parameter for promise

kleekich21 committed:
test: add tests

kleekich21 committed:
refactor: manages local storages in different directory

kleekich21 committed:
refactor: change path for local storage

kleekich21 committed:
test: test wip
